src/worker/queue.ts

```diff
@@ -48,3 +51,66 @@ export function startQueue(
 
     return worker
 }
+
+function startQueueKafka(
+    server: PluginsServer,
+    processEvent: (event: PluginEvent) => Promise<PluginEvent | null>
+): Queue {
+    const eventsProcessor = new EventsProcessor(server)
+
+    eventsProcessor.kafkaConsumer.on('ready', () => {
+        eventsProcessor.kafkaConsumer.subscribe([KAFKA_EVENTS_HANDOFF])
+        // consume event messages in batches of 100 every 100 ms
+        setInterval(() => {
```

Comment:
This limits us to max 1000 messages/sec, which will not be enough for app... and is way below what the plugin server could handle. So we need something more robust here. 

Celery uses [something like this](https://github.com/PostHog/posthog-plugin-server/blob/4037c91f8b64083ef6a99f7ff34bab8cfdb332ab/src/celery/broker.ts#L104), where it just asks for the next message when it can.

The way `piscina` currently works is that we spawn `WORKER_CONCURRENCY` (e.g. 8) threads that can handle `TASKS_PER_WORKER` (defaults to 10) async tasks. When 8 * 10 threads are being processed, the rest get put into piscina's queue.

Currently if celery produces tasks fast enough and piscina remains busy, the queue will grow endlessly. This is a problem since if the server then crashes (e.g. out of memory), we lose all the events in the queue. The correct approach would be to only ask celery for another task if we have a worker that can handle it. Basically in this example if there are more than 80 promises pending with piscina, we shouldn't ask for a new task with celery (instead of the current `process.nextTick` logic).

The same logic should be applied to receiving a batch of messages from Kafka as well.

There's [some piscina code here](https://github.com/piscinajs/piscina#backpressure) regarding maximum queue sizes that could help, but AFAIK that code waits until it has processed all events (queue 0 and all workers idle) before then asking for the next event from that stream, so it's not the full solution. We should always keep our workers busy.

Reply:
100% agree that 1k events/s is too low. Well, we can just consume messages one at a time as they come ([node-rdkafka "flowing mode"](https://github.com/Blizzard/node-rdkafka#standard-api-1)), but then batching is more work on our side (and possibly slightly less performant). We can also up batch size and consumption frequency to some arbitrary amounts e.g. 10k events every 50 ms (so 20k events/s). I believe batching is something we're not actually using yet (not in the Celery pipeline and also not here), but the aim is to make use of it.